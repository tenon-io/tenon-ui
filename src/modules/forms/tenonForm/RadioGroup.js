import React, { Fragment, forwardRef } from 'react';
import classNames from 'classnames';
import PropTypes from 'prop-types';
import FocusCatcher from '../../utils/components/FocusCatcher';
import RequiredLegend from '../RequiredLegend';
import FeedbackBlock from './FeedbackBlock';

/**
 * @component
 * A default view component to use with the tenon-ui
 * smart radiogroup controller.
 *
 * @prop required {string} legend - Text to show as the legend.
 * @prop required {object} options - Key mapping object to show
 *                the radiobuttons from. The keys are used as
 *                radiobutton values, the values as the labels.
 * @prop {string} contentHintText - Text to show as content
 *                hint of the radiogroup.
 * @prop {string} errorText - Text to show as error message
 *                of the radiogroup.
 * @prop required {function} getLabelProps - Prop getter function
 *              from the smart radiogroup controller for each <label>
 *              of the group.
 * @prop required {function} getLegendProps - Prop getter function
 *              from the smart radiogroup controller for the <legend>
 *              of the group.
 * @prop required {function} getRadioButtonProps - Prop getter function
 *              from the smart radiogroup controller for each <input>
 *              of the group.
 * @prop required {function} getRadioGroupProps - Prop getter function
 *              from the smart radiogroup controller for the grouping
 *              element carrying the 'radiogroup' role.
 * @prop {function} getErrorProps - Prop getter function
 *              from the smart radiogroup controller for the
 *              error message container.
 * @prop {function} getContentHintProps - Prop getter
 *              function from the smart radiogroup controller
 *              for the content hint container.
 * @prop {boolean} showError - A flag to indicate when the
 *              error container should be displayed.
 * @prop {string} required - Set to 'required' to mark the
 *              radiogroup container as required.
 * @prop {string} className - An optional class string to transfer to the
 *              className prop of the select element.
 * @prop {string} requiredText - An optional text to display next to the
 *              legend if the group is required.
 */
const RadioGroup = forwardRef(
    (
        {
            legend,
            options,
            getRadioButtonProps,
            getLabelProps,
            getLegendProps,
            getRadioGroupProps,
            contentHintText,
            errorText,
            getErrorProps,
            getContentHintProps,
            showError,
            required,
            requiredText,
            className
        },
        ref
    ) => (
        <fieldset className={classNames('form-group', className)}>
            <RequiredLegend
                requiredText={requiredText || null}
                {...getLegendProps()}
            >
                {legend}
            </RequiredLegend>
            <div
                ref={ref}
                {...getRadioGroupProps({
                    required: required || null
                })}
            >
                <FocusCatcher>
                    {Object.keys(options).map(option => (
                        <Fragment key={option}>
                            <input
                                {...getRadioButtonProps({
                                    value: option
                                })}
                            />
                            <label
                                {...getLabelProps({
                                    autoIdPostfix: option
                                })}
                            >
                                {options[option]}
                            </label>
                        </Fragment>
                    ))}
                </FocusCatcher>
            </div>
            <FeedbackBlock
                getContentHintProps={getContentHintProps}
                getErrorProps={getErrorProps}
                errorText={errorText}
                contentHintText={contentHintText}
                showError={showError}
            />
        </fieldset>
    )
);

RadioGroup.displayName = 'RadioGroup';

RadioGroup.propTypes = {
    legend: PropTypes.string.isRequired,
    options: PropTypes.object.isRequired,
    getRadioButtonProps: PropTypes.func.isRequired,
    getLegendProps: PropTypes.func.isRequired,
    getRadioGroupProps: PropTypes.func.isRequired,
    getLabelProps: PropTypes.func.isRequired,
    getErrorProps: PropTypes.func,
    getContentHintProps: PropTypes.func,
    required: PropTypes.string,
    contentHintText: PropTypes.string,
    errorText: PropTypes.string,
    showError: PropTypes.bool,
    className: PropTypes.string,
    requiredText: PropTypes.string
};

export default RadioGroup;
