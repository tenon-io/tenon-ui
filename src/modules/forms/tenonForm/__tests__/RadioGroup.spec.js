jest.mock('../../../utils/components/FocusCatcher');

import React, { createRef } from 'react';
import { render, cleanup } from 'react-testing-library';
import 'jest-dom/extend-expect';
import RadioGroup from '../RadioGroup';

describe('RadioGroup', () => {
    afterEach(cleanup);

    it('should render an input and label only', () => {
        const { container, getByLabelText } = render(
            <RadioGroup
                legend="Test radio group"
                options={{
                    one: 'One',
                    two: 'Two'
                }}
                getLegendProps={() => ({ id: 'bar' })}
                getLabelProps={props => ({
                    htmlFor: `foo-${props.autoIdPostfix}`
                })}
                getRadioButtonProps={({ value }) => ({
                    id: `foo-${value}`,
                    value,
                    onChange: jest.fn()
                })}
                getRadioGroupProps={conf => ({
                    'aria-labelledby': 'bar',
                    role: 'radiogroup',
                    value: conf.value
                })}
            />
        );

        expect(getByLabelText('One').value).toBe('one');

        expect(container.firstChild).toMatchSnapshot();
    });

    it('should render an error text when invalid', () => {
        const { container } = render(
            <RadioGroup
                legend="Test radio group"
                options={{
                    one: 'One',
                    two: 'Two'
                }}
                getLegendProps={() => ({ id: 'bar' })}
                getLabelProps={props => ({
                    htmlFor: `foo-${props.autoIdPostfix}`
                })}
                getRadioButtonProps={({ value }) => ({
                    id: `foo-${value}`,
                    value,
                    onChange: jest.fn()
                })}
                getRadioGroupProps={conf => ({
                    'aria-labelledby': 'bar',
                    'aria-describedby': 'err',
                    role: 'radiogroup',
                    value: conf.value
                })}
                getErrorProps={() => ({ id: 'err' })}
                errorText="Test error"
                showError={true}
            />
        );

        expect(container.firstChild).toMatchSnapshot();
    });

    it('should render a content hint if given', () => {
        const { container } = render(
            <RadioGroup
                legend="Test radio group"
                options={{
                    one: 'One',
                    two: 'Two'
                }}
                getLegendProps={() => ({ id: 'bar' })}
                getLabelProps={props => ({
                    htmlFor: `foo-${props.autoIdPostfix}`
                })}
                getRadioButtonProps={({ value }) => ({
                    id: `foo-${value}`,
                    value,
                    onChange: jest.fn()
                })}
                getRadioGroupProps={conf => ({
                    'aria-labelledby': 'bar',
                    'aria-describedby': 'inf',
                    role: 'radiogroup',
                    value: conf.value
                })}
                getContentHintProps={() => ({ id: 'inf' })}
                contentHintText="Content hint test"
            />
        );

        expect(container.firstChild).toMatchSnapshot();
    });

    it('should allow easy focus of the internal radio group', () => {
        const radioGroupRef = createRef();

        const { container } = render(
            <RadioGroup
                ref={radioGroupRef}
                legend="Test radio group"
                options={{
                    one: 'One',
                    two: 'Two'
                }}
                getLegendProps={() => ({ id: 'bar' })}
                getLabelProps={props => ({
                    htmlFor: `foo-${props.autoIdPostfix}`
                })}
                getRadioButtonProps={({ value }) => ({
                    id: `foo-${value}`,
                    value,
                    onChange: jest.fn()
                })}
                getRadioGroupProps={conf => ({
                    'aria-labelledby': 'bar',
                    role: 'radiogroup',
                    tabIndex: '-1',
                    value: conf.value
                })}
            />
        );
        expect(container.querySelector('[role="radiogroup"]')).toEqual(
            radioGroupRef.current
        );
    });
});
