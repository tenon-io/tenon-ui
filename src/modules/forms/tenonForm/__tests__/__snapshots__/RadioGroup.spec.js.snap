// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RadioGroup should render a content hint if given 1`] = `
<fieldset
  class="form-group"
>
  <legend
    id="bar"
  >
    Test radio group
  </legend>
  <div
    aria-describedby="inf"
    aria-labelledby="bar"
    role="radiogroup"
  >
    <div>
      <span>
        FocusCatcher
      </span>
      <input
        id="foo-one"
        value="one"
      />
      <label
        for="foo-one"
      >
        One
      </label>
      <input
        id="foo-two"
        value="two"
      />
      <label
        for="foo-two"
      >
        Two
      </label>
    </div>
  </div>
  <div
    class="info-wrapper"
  >
    <span
      class="icon"
    />
    <span
      id="inf"
    >
      Content hint test
    </span>
  </div>
</fieldset>
`;

exports[`RadioGroup should render an error text when invalid 1`] = `
<fieldset
  class="form-group"
>
  <legend
    id="bar"
  >
    Test radio group
  </legend>
  <div
    aria-describedby="err"
    aria-labelledby="bar"
    role="radiogroup"
  >
    <div>
      <span>
        FocusCatcher
      </span>
      <input
        id="foo-one"
        value="one"
      />
      <label
        for="foo-one"
      >
        One
      </label>
      <input
        id="foo-two"
        value="two"
      />
      <label
        for="foo-two"
      >
        Two
      </label>
    </div>
  </div>
  <div
    class="error-wrapper"
  >
    <span
      class="icon"
    />
    <span
      id="err"
    >
      Test error
    </span>
  </div>
</fieldset>
`;

exports[`RadioGroup should render an input and label only 1`] = `
<fieldset
  class="form-group"
>
  <legend
    id="bar"
  >
    Test radio group
  </legend>
  <div
    aria-labelledby="bar"
    role="radiogroup"
  >
    <div>
      <span>
        FocusCatcher
      </span>
      <input
        id="foo-one"
        value="one"
      />
      <label
        for="foo-one"
      >
        One
      </label>
      <input
        id="foo-two"
        value="two"
      />
      <label
        for="foo-two"
      >
        Two
      </label>
    </div>
  </div>
</fieldset>
`;
